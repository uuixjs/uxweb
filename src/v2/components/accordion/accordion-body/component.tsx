import { Background, BorderRadiusProps, Display, Layout } from "../../layout";
import { FC, ReactNode } from "react";

import { getAriaProps } from "lib";

export interface AccordionBodyProps {
  /**
   * @example Lorem ipsum dolor sit amet, consectetur adipiscing elit
   */
  children?: ReactNode;
  /**
   * @example true
   */
  isOpen?: boolean;
  borderRadius?: BorderRadiusProps;
}

/**
 * AccordionBody is a content region whose visibility should be
 * controlled by an AccordionHeader
 */
export const AccordionBody: FC<AccordionBodyProps> = (props) => (
  <Layout
    padding={2}
    background={Background.Base}
    elevation={1}
    display={props.isOpen ? Display.Block : Display.Hide}
    borderRadius={props.borderRadius}
    {...getAriaProps(props)}
  >
    {props.children}
  </Layout>
);

AccordionBody.displayName = "AccordionBody";
