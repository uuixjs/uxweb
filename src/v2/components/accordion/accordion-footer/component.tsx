import { Button, ButtonProps, ButtonType } from "../../button/button";
import { Display, JustifyContent, Layout } from "../../layout";

import { FC } from "react";
import { getAriaProps } from "lib/ui-utils";

export interface AccordionFooterProps {
  primaryButtonProps: ButtonProps;
  secondaryButtonProps?: ButtonProps;
}

/**
 * AccordionFooter should be placed within an AccordionBody and
 * composes either one or two call-to-action buttons with consistent placement.
 */
export const AccordionFooter: FC<AccordionFooterProps> = (props) => {
  const primaryButton = (
    // Order of attributes matters; later props will override earlier props
    <Button {...props.primaryButtonProps} variant={ButtonType.Primary} />
  );

  let secondaryButton;

  if (props.secondaryButtonProps) {
    secondaryButton = (
      // Order of attributes matters; later props will override earlier props
      <Button {...props.secondaryButtonProps} variant={ButtonType.Secondary} />
    );
  }

  return (
    <Layout
      display={Display.Flex}
      justifyContent={JustifyContent.End}
      margin={{ top: 2 }}
      {...getAriaProps(props)}
    >
      {secondaryButton && (
        <Layout margin={{ left: 1 }}>{secondaryButton}</Layout>
      )}

      <Layout margin={{ left: 1 }}>{primaryButton}</Layout>
    </Layout>
  );
};

AccordionFooter.displayName = "AccordionFooter";
