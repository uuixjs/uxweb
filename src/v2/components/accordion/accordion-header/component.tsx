import { Align, Aspect, AspectRatio } from "../../aspect";
import {
  AlignItems,
  Background,
  BorderRadiusProps,
  Color,
  Display,
  FlexDirection,
  InjectLayout,
  JustifyContent,
  Layout,
  Overflow,
  Position,
  ZIndex,
} from "../../layout";
import { Component, MouseEvent } from "react";
import { CoreImage, CoreImageProps } from "../../core-image";
import { CoreText, TextType } from "../../core-text";
import {
  ExpandableIndicator,
  ExpandableIndicatorDirection,
} from "../../expandable-indicator";
import { PaddingValue, getAriaProps, styled } from "lib";
import { Title, TitleSize } from "../../title";

import { Interactable } from "../../interactable";

export enum AccordionHeaderSize {
  Default = "default",
  Narrow = "narrow",
}

export interface AccordionDefaultHeaderProps {
  index?: number;
  /**
   * @example What is thy bidding?
   */
  title: string;
  /**
   * @example Reach 10 followers
   */
  description?: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  imageProps?: CoreImageProps;
  imageRatio?: AspectRatio.Aspect1x1 | AspectRatio.Aspect3x4;
  backgroundImageProps?: CoreImageProps;
  imageLabelOverlay?: string;
  label?: string;
  borderRadius?: BorderRadiusProps;
  size?: AccordionHeaderSize.Default;
}

export interface AccordionNarrowHeaderProps {
  index?: number;
  title: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  borderRadius?: BorderRadiusProps;
  size: AccordionHeaderSize.Narrow;
}

const defaultHeight = "8rem";
const narrowHeight = "5rem";

const ScAccordionHeaderContents = styled(Layout)<{
  $size: AccordionHeaderSize;
}>(({ $size }) =>
  $size && $size === AccordionHeaderSize.Default
    ? `min-height: ${defaultHeight};`
    : `height: ${narrowHeight};`,
);

const ScAccordionHeaderImageContainer = styled(Layout)<{
  $ratio: AccordionDefaultHeaderProps["imageRatio"];
}>(({ $ratio }) =>
  $ratio && $ratio === AspectRatio.Aspect3x4
    ? `min-width: calc(${defaultHeight} * (3/4));`
    : `min-width: ${defaultHeight};`,
);

const ScAccordionHeaderBackgroundImage = styled(CoreImage)`
  opacity: 0.5;
  object-fit: cover;
  object-position: center;
`;

export type AccordionHeaderProps =
  | AccordionDefaultHeaderProps
  | AccordionNarrowHeaderProps;

/**
 * An AccordionHeader is an interactable region which should be used to toggle
 * the visibility of additional information.
 */
export class AccordionHeader extends Component<AccordionHeaderProps, {}> {
  public render() {
    return (
      <Layout
        className="accordion-header"
        background={Background.Base}
        elevation={1}
        position={Position.Relative}
        borderRadius={this.props.borderRadius}
        {...getAriaProps(this.props)}
      >
        {this.renderBackgroundImage()}

        <Interactable
          onClick={this.onClick}
          aria-expanded={this.props.isOpen}
          borderRadius={this.props.borderRadius}
        >
          <ScAccordionHeaderContents
            $size={
              this.props.size === AccordionHeaderSize.Default ||
              this.props.size === undefined
                ? AccordionHeaderSize.Default
                : AccordionHeaderSize.Narrow
            }
            display={Display.Flex}
            position={Position.Relative}
            borderRadius={this.props.borderRadius}
            overflow={Overflow.Hidden}
          >
            {this.renderImageAndLabel()}
            {this.renderTitles()}
            {this.renderLabel()}
            {this.renderCaret()}
          </ScAccordionHeaderContents>
        </Interactable>
      </Layout>
    );
  }

  private getYPadding = (): PaddingValue => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return 2;
    }
    return this.props.description ? 2 : 3;
  };

  private renderCaret = () => {
    return (
      <Layout
        flexGrow={0}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.Center}
        padding={{ y: this.getYPadding(), right: 2 }}
      >
        <ExpandableIndicator
          open={this.props.isOpen}
          openDirection={ExpandableIndicatorDirection.Up}
          closedDirection={ExpandableIndicatorDirection.Down}
        />
      </Layout>
    );
  };

  private renderTitles = () => {
    return (
      <Layout
        flexGrow={1}
        flexShrink={1}
        display={Display.Flex}
        flexDirection={FlexDirection.Column}
        alignItems={AlignItems.Start}
        justifyContent={JustifyContent.Center}
        overflow={Overflow.Hidden}
        padding={{ y: this.getYPadding(), left: 2, right: 1 }}
      >
        <Layout fullWidth>
          <Title type={TextType.H3} size={TitleSize.Small} ellipsis>
            {this.props.title}
          </Title>
          {!isAccordionHeaderNarrowProps(this.props) && this.props.description && (
            <CoreText color={Color.Alt2} ellipsis>
              {this.props.description}
            </CoreText>
          )}
        </Layout>
      </Layout>
    );
  };

  private renderBackgroundImage = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.backgroundImageProps) {
      return;
    }

    return (
      <InjectLayout
        className="accordion-header__background-image"
        position={Position.Absolute}
        attachTop
        attachLeft
        fullWidth
        fullHeight
        borderRadius={this.props.borderRadius}
      >
        <ScAccordionHeaderBackgroundImage
          {...this.props.backgroundImageProps}
        />
      </InjectLayout>
    );
  };

  private renderLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.label) {
      return;
    }

    return (
      <Layout
        padding={{ right: 2 }}
        flexGrow={1}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.End}
      >
        <CoreText>{this.props.label}</CoreText>
      </Layout>
    );
  };

  private renderImageAndLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.imageProps && !this.props.imageLabelOverlay) {
      return;
    }

    return (
      <ScAccordionHeaderImageContainer
        $ratio={this.props.imageRatio}
        flexGrow={0}
        flexShrink={0}
        background={Background.Alt2}
        position={Position.Relative}
      >
        {this.props.imageLabelOverlay && (
          <Layout
            position={Position.Relative}
            fullHeight
            zIndex={ZIndex.Above}
            display={Display.Flex}
            alignItems={AlignItems.Center}
            justifyContent={JustifyContent.Center}
            padding={1}
            color={this.props.imageProps ? Color.Overlay : Color.Inherit}
            background={this.props.imageProps ? Background.Overlay : undefined}
          >
            <Title>{this.props.imageLabelOverlay}</Title>
          </Layout>
        )}
        {this.props.imageProps && (
          <Layout
            position={Position.Absolute}
            fullWidth
            fullHeight
            attachTop
            attachLeft
            overflow={Overflow.Hidden}
          >
            <Aspect
              ratio={this.props.imageRatio || AspectRatio.Aspect1x1}
              align={Align.Center}
            >
              <CoreImage {...this.props.imageProps} />
            </Aspect>
          </Layout>
        )}
      </ScAccordionHeaderImageContainer>
    );
  };

  private onClick = (event: MouseEvent<HTMLElement>) => {
    if (this.props.onClick !== undefined) {
      this.props.onClick(event, this.props.index);
    }
  };
}

function isAccordionHeaderNarrowProps(
  props: AccordionDefaultHeaderProps | AccordionNarrowHeaderProps,
): props is AccordionNarrowHeaderProps {
  return props.size === AccordionHeaderSize.Narrow;
}
