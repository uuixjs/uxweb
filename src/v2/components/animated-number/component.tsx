import { Component } from "react";
import CountUp from "react-countup";
import { styled } from "lib/ui-utils";

export interface AnimatedNumberProps {
  value: number;
  formatValue: (n: number) => string;
  initialValue?: number;
  duration?: number;
  children?: undefined;
  monospaced?: boolean;
}

export interface AnimatedNumberState {
  prevValue: number;
  currentValue: number;
}

export const ScAnimatedNumber = styled(CountUp)<{
  $monospaced: AnimatedNumberProps["monospaced"];
}>(({ $monospaced }) => {
  return (
    $monospaced && {
      "font-feature-settings": `"tnum"`,
    }
  );
});
export class AnimatedNumber extends Component<
  AnimatedNumberProps,
  AnimatedNumberState
> {
  public static defaultProps: Partial<AnimatedNumberProps> = {
    duration: 0.7,
    monospaced: true,
  };

  public constructor(props: AnimatedNumberProps) {
    super(props);
    this.state = {
      currentValue: props.value,
      prevValue:
        props.initialValue !== undefined ? props.initialValue : props.value,
    };
  }

  public componentDidUpdate() {
    if (this.props.value !== this.state.currentValue) {
      this.setState((s) => ({
        currentValue: this.props.value,
        prevValue: s.currentValue,
      }));
    }
  }

  public render() {
    const {
      duration,
      formatValue,
      initialValue,
      monospaced,
      value,
      ...props
    } = this.props;
    return (
      <ScAnimatedNumber
        $monospaced={monospaced}
        duration={duration}
        end={this.state.currentValue}
        start={this.state.prevValue}
        formattingFn={formatValue}
        {...props}
      />
    );
  }
}
