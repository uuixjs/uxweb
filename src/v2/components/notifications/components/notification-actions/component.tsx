import { FC, ReactNode } from "react";
import { Button, ButtonSize, ButtonType } from "../../../button/button";
import { CoreButtonSize } from "../../../button/core-button";
import { StatusButton } from "../../../button/status-button";
import { CoreInteractivePublicProps } from "../../../core-interactive";
import {
  Display,
  FlexDirection,
  JustifyContent,
  Layout,
} from "../../../layout";
import { SVGAsset } from "../../../svg";
import { NotificationType } from "../../notification";

export interface NotificationAction extends CoreInteractivePublicProps {
  children?: ReactNode;
  icon?: SVGAsset;
  overlay?: boolean;
  dropdown?: boolean;
  type?: ButtonType;
}

export interface NotificationActionsProps {
  primaryButton: NotificationAction;
  secondaryButton?: NotificationAction;
  type?: NotificationType;
}

export const NotificationActions: FC<NotificationActionsProps> = ({
  type,
  primaryButton,
  secondaryButton,
}) => {
  return (
    <Layout
      display={Display.Flex}
      justifyContent={JustifyContent.Start}
      flexDirection={FlexDirection.RowReverse}
    >
      <div>
        {type && !primaryButton.type ? (
          <StatusButton
            size={CoreButtonSize.Small}
            {...primaryButton}
            type={type}
          />
        ) : (
          <Button
            size={ButtonSize.Small}
            variant={primaryButton.type || ButtonType.Primary}
            {...primaryButton}
            type={undefined}
          />
        )}
      </div>
      {secondaryButton && (
        <Layout margin={{ right: 1 }}>
          <Button
            size={ButtonSize.Small}
            variant={secondaryButton.type || ButtonType.Secondary}
            {...secondaryButton}
            type={undefined}
          />
        </Layout>
      )}
    </Layout>
  );
};

NotificationActions.displayName = "NotificationActions";
