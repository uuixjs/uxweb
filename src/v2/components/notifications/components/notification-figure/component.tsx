import { Avatar, AvatarProps } from "../../../avatar";
import { styleVariant, styled, themeTokenRule } from "lib/ui-utils";

import { FC } from "react";
import { Icon } from "../../../icon";
import { NotificationType } from "../../notification";
import { SVGAsset } from "../../../svg";

export interface NotificationFigureProps {
  /** Optional avatar to replace the default SVG icon. */
  avatar?: Pick<AvatarProps, Exclude<keyof AvatarProps, "size">>;
  /**
   * The type of notification.
   *
   * @example Info
   */
  type: NotificationType;
}

const NOTIFICATION_ICON: { [key: string]: SVGAsset } = {
  [NotificationType.Error]: SVGAsset.NotificationError,
  [NotificationType.Info]: SVGAsset.NotificationInfo,
  [NotificationType.Success]: SVGAsset.NotificationSuccess,
  [NotificationType.Warning]: SVGAsset.NotificationWarning,
};

const ScNotificationFigure = styled.div<{
  type: NotificationFigureProps["type"];
}>`
  display: flex;
  align-items: center;
  color: ${styleVariant("type", {
    [NotificationType.Error]: themeTokenRule("color-fill-alert"),
    [NotificationType.Info]: themeTokenRule("color-fill-info"),
    [NotificationType.Success]: themeTokenRule("color-fill-success"),
    [NotificationType.Warning]: themeTokenRule("color-fill-warn"),
  })};
`;

export const NotificationFigure: FC<NotificationFigureProps> = ({
  avatar,
  ...props
}) => {
  return (
    <ScNotificationFigure className="tw-notification-figure" {...props}>
      {avatar ? (
        <Avatar size={30} {...avatar} />
      ) : (
        <Icon asset={NOTIFICATION_ICON[props.type]} />
      )}
    </ScNotificationFigure>
  );
};

NotificationFigure.displayName = "NotificationFigure";
