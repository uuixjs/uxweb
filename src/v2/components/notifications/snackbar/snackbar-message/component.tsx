import { AlignItems, Display, Layout, Position } from "../../../layout";
import { LineHeight, TextType } from "../../../core-text";
import { Title, TitleSize } from "../../../title";
import { getAriaProps, styled } from "lib";

import { Component } from "react";
import { Icon } from "../../../icon";
import { NotificationType } from "../../notification";
import { SVGAsset } from "../../../svg";

export interface SnackbarMessageProps {
  iconType?: NotificationType;
  title?: string | JSX.Element;
}

const SNACKBAR_ICON: { [key: string]: SVGAsset } = {
  [NotificationType.Error]: SVGAsset.NotificationError,
  [NotificationType.Info]: SVGAsset.NotificationInfo,
  [NotificationType.Success]: SVGAsset.NotificationSuccess,
  [NotificationType.Warning]: SVGAsset.NotificationWarning,
};

const ScSnackbarMessage = styled.div`
  display: inline-flex;
  align-items: center;
  min-height: 3rem;
`;

export class SnackbarMessage extends Component<SnackbarMessageProps> {
  public render() {
    const icon = this.props.iconType && (
      <Layout margin={{ x: 0.5 }}>
        <Icon asset={SNACKBAR_ICON[this.props.iconType]} />
      </Layout>
    );

    const title =
      this.props.title && typeof this.props.title === "string" ? (
        <Layout flexGrow={1} margin={{ left: 0.5, right: 1 }}>
          <Title
            lineHeight={LineHeight.Heading}
            type={TextType.H5}
            size={TitleSize.ExtraSmall}
          >
            {this.props.title}
          </Title>
        </Layout>
      ) : (
        this.props.title
      );

    return (
      <Layout {...getAriaProps(this.props)}>
        {(icon || title) && (
          <ScSnackbarMessage className="tw-snackbar-message">
            {icon && (
              <Layout
                margin={{ top: 0.5 }}
                display={Display.Flex}
                position={Position.Relative}
                alignItems={AlignItems.Start}
                flexGrow={0}
                flexShrink={0}
              >
                {icon}
              </Layout>
            )}
            {title}
          </ScSnackbarMessage>
        )}
      </Layout>
    );
  }
}
