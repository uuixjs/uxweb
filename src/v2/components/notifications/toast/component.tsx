import {
  AlignItems,
  Background,
  Display,
  FlexWrap,
  JustifyContent,
  Layout,
} from "../../layout";
import { FC, MouseEventHandler, ReactNode } from "react";
import { staticTokenRule, styled } from "lib/ui-utils";

import { Icon } from "../../icon";
import { Interactable } from "../../interactable";
import { SVGAsset } from "../../svg";

export interface ToastProps {
  children?: ReactNode;
  tabIndex?: number;
  /**
   * @example ()=>{ alert('Close Event Detected') }
   */
  onClose: MouseEventHandler<HTMLElement>;
}

const ScToast = styled(Layout)`
  width: 100%;
  min-width: 30rem;
  z-index: ${staticTokenRule("z-index-notification")};
`;

const ScToastDismiss = styled(Layout)`
  /* Google minimum recommended touch target size.
  https://developers.google.com/speed/docs/insights/SizeTapTargetsAppropriately */
  min-width: 4.8rem;
  min-height: 4.8rem;
`;

// @deprecated Consider the Snackbar pattern instead.
export const Toast: FC<ToastProps> = (props) => {
  return (
    <Layout fullWidth>
      <ScToast
        display={Display.Flex}
        flexWrap={FlexWrap.NoWrap}
        className="tw-toast"
        elevation={3}
        background={Background.Base}
        border
      >
        <Layout flexGrow={1}>{props.children}</Layout>
        <Layout display={Display.Flex} flexShrink={0} borderLeft>
          <Interactable
            onClick={props.onClose}
            aria-label="Dismiss This"
            tabIndex={props.tabIndex}
            data-a-target="tw-toast-close"
          >
            <ScToastDismiss
              padding={1}
              display={Display.Flex}
              alignItems={AlignItems.Center}
              justifyContent={JustifyContent.Center}
              className="tw-toast__dismiss-target"
            >
              <Icon asset={SVGAsset.Close} />
            </ScToastDismiss>
          </Interactable>
        </Layout>
      </ScToast>
    </Layout>
  );
};

Toast.displayName = "Toast";
