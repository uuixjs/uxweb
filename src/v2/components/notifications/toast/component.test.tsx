import { setupShallowTest } from "../../../tests/helpers";
import { Toast, ToastProps } from "./component";

function getOptionalProps(): ToastProps {
  return {
    tabIndex: 0,
    onClose: jest.fn(),
  };
}

const setupShallow = setupShallowTest(Toast);

describe("Toast", () => {
  it("renders toast component with only required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders toast component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });
});
