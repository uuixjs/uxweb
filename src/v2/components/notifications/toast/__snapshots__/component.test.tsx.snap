// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders toast component with all props 1`] = `
<Layout
  fullWidth={true}
>
  <Styled(Layout)
    background="base"
    border={true}
    className="tw-toast"
    display="flex"
    elevation={3}
    flexWrap="nowrap"
  >
    <Layout
      flexGrow={1}
    />
    <Layout
      borderLeft={true}
      display="flex"
      flexShrink={0}
    >
      <ForwardRef(Interactable)
        aria-label="Dismiss This"
        data-a-target="tw-toast-close"
        onClick={[MockFunction]}
        tabIndex={0}
      >
        <Styled(Layout)
          alignItems="center"
          className="tw-toast__dismiss-target"
          display="flex"
          justifyContent="center"
          padding={1}
        >
          <Icon
            asset="Close"
          />
        </Styled(Layout)>
      </ForwardRef(Interactable)>
    </Layout>
  </Styled(Layout)>
</Layout>
`;

exports[`Toast renders toast component with only required props 1`] = `
<Layout
  fullWidth={true}
>
  <Styled(Layout)
    background="base"
    border={true}
    className="tw-toast"
    display="flex"
    elevation={3}
    flexWrap="nowrap"
  >
    <Layout
      flexGrow={1}
    />
    <Layout
      borderLeft={true}
      display="flex"
      flexShrink={0}
    >
      <ForwardRef(Interactable)
        aria-label="Dismiss This"
        data-a-target="tw-toast-close"
      >
        <Styled(Layout)
          alignItems="center"
          className="tw-toast__dismiss-target"
          display="flex"
          justifyContent="center"
          padding={1}
        >
          <Icon
            asset="Close"
          />
        </Styled(Layout)>
      </ForwardRef(Interactable)>
    </Layout>
  </Styled(Layout)>
</Layout>
`;
